File: Algorithms/p002_Add_Two_Numbers/p002_Add_Two_Numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head = ListNode(None)
        tail = head
        res = 0
        while l1 or l2 or res:
            if l1:
                res += l1.val
                l1 = l1.next
            if l2:
                res += l2.val
                l2 = l2.next
            tail.next = ListNode(res%10)
            tail = tail.next
            res //= 10
        return head.next

File: Algorithms/p002_Add_Two_Numbers/p002_Add_Two_Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode head(0);
        ListNode* tail = &head;
        int carry = 0;
        while(l1||l2||carry){
            int op1 = l1?l1->val:0;
            int op2 = l2?l2->val:0;
            int res = op1+op2+carry;
            carry = (res>=10)?1:0;
            if(res>=10) res -= 10;
            tail->next = new ListNode(res);
            tail = tail->next;
            if(l1) l1 = l1->next;
            if(l2) l2 = l2->next;
        }
        return head.next;
    }
};
